File: frontend/src/types/scenario.ts
// 场景编辑器用的内部数据结构(区别于后端 DTO)
// 编辑器需要保存完整的 Config,DTO 只是简化的列表展示
export interface RequestEditor {
  uid: string                  // 编辑器内部的唯一 ID,用于 React key
  name: string
  method: string
  url: string
  headers: HeaderEntry[]
  bodyType: 'json' | 'form' | 'raw' | 'none'
  bodyJson: string
  bodyRaw: string
  extractors: ExtractorEntry[]
  assertions: AssertionEntry
  thinkTime: number             // ms,0 = 无
  weight: number               // 用于多请求时的权重
}

export interface HeaderEntry {
  uid: string
  key: string
  value: string
  enabled: boolean
}

export interface ExtractorEntry {
  uid: string
  name: string                  // 变量名
  jsonPath: string              // $.data.token
  enabled: boolean
}

export interface AssertionEntry {
  status: number[]              // [200, 201]
  maxLatencyMs: number          // P95 上限, 0 = 不检查
  bodyContains: string[]        // body 必须包含的字符串
  enabled: boolean
}

export interface VariableEntry {
  uid: string
  name: string                  // 变量名(用作 {{name}} 占位符)
  values: string[]              // 多个值 → 数据驱动
  enabled: boolean
}

export interface ThresholdConfig {
  p95Ms: number                 // P95 延迟上限 (ms), 0 = 不检查
  p99Ms: number
  errorRate: number             // 0-1
  minRPS: number                // 0 = 不检查
}

export interface LoadConfig {
  vus: number
  duration: string              // "5m", "30s", "1h"
  rampUpType: 'linear' | 'step' | 'wave' | 'none'
  rampUpDuration: string
  targetRPS: number             // 0 = 不限
  thinkTimeMs: number           // 默认 think time (ms)
}

export interface ScenarioEditorData {
  uid: string                   // 编辑器 ID,= "new" 或 scenario id
  name: string
  description: string
  requests: RequestEditor[]
  variables: VariableEntry[]
  thresholds: ThresholdConfig
  load: LoadConfig
  tags: string[]
}

export function createEmptyRequest(name = ''): RequestEditor {
  return {
    uid: generateUid(),
    name: name || 'New Request',
    method: 'GET',
    url: '/api/...',
    headers: [],
    bodyType: 'none',
    bodyJson: '{}',
    bodyRaw: '',
    extractors: [],
    assertions: {
      status: [200],
      maxLatencyMs: 0,
      bodyContains: [],
      enabled: true,
    },
    thinkTime: 0,
    weight: 1,
  }
}

export function createEmptyVariable(): VariableEntry {
  return {
    uid: generateUid(),
    name: '',
    values: [''],
    enabled: true,
  }
}

let _uidCounter = 0
export function generateUid(): string {
  _uidCounter++
  return `e_${Date.now()}_${_uidCounter}_${Math.random().toString(36).slice(2, 7)}`
}

// 转换为后端 DTO 的 Config
export function toScenarioConfig(data: ScenarioEditorData): any {
  return {
    load: {
      vus: data.load.vus,
      duration: data.load.duration,
      targetRPS: data.load.targetRPS || 0,
      rampUp: data.load.rampUpType !== 'none' ? {
        type: data.load.rampUpType,
        duration: data.load.rampUpDuration || '30s',
      } : undefined,
      thinkTime: msToDuration(data.load.thinkTimeMs),
    },
    requests: data.requests.map(r => ({
      name: r.name,
      method: r.method,
      url: r.url,
      headers: headersToObject(r.headers),
      body: r.bodyType === 'json' ? safeParseJSON(r.bodyJson) : undefined,
      bodyRaw: r.bodyType === 'raw' ? r.bodyRaw : undefined,
      extractors: extractorsToObject(r.extractors),
      assertions: r.assertions.enabled ? {
        status: r.assertions.status,
        maxLatencyMs: r.assertions.maxLatencyMs,
        bodyContains: r.assertions.bodyContains.filter(s => s.trim()),
      } : undefined,
      thinkTime: msToDuration(r.thinkTime),
      weight: r.weight,
    })),
    variables: variablesToObject(data.variables),
    thresholds: {
      p95Ms: data.thresholds.p95Ms,
      p99Ms: data.thresholds.p99Ms,
      errorRate: data.thresholds.errorRate,
      minRPS: data.thresholds.minRPS,
    },
  }
}

function headersToObject(headers: HeaderEntry[]): Record<string, string> {
  const obj: Record<string, string> = {}
  for (const h of headers) {
    if (h.enabled && h.key.trim()) {
      obj[h.key.trim()] = h.value
    }
  }
  return obj
}

function extractorsToObject(extractors: ExtractorEntry[]): Record<string, string> {
  const obj: Record<string, string> = {}
  for (const e of extractors) {
    if (e.enabled && e.name.trim() && e.jsonPath.trim()) {
      obj[e.name.trim()] = e.jsonPath.trim()
    }
  }
  return obj
}

function variablesToObject(vars: VariableEntry[]): Record<string, any> {
  const obj: Record<string, any> = {}
  for (const v of vars) {
    if (v.enabled && v.name.trim()) {
      obj[v.name.trim()] = v.values.length > 1 ? v.values : v.values[0]
    }
  }
  return obj
}

function safeParseJSON(text: string): any {
  if (!text.trim()) return undefined
  try {
    return JSON.parse(text)
  } catch {
    return undefined
  }
}

function msToDuration(ms: number): string {
  if (ms <= 0) return ''
  if (ms < 1000) return `${ms}ms`
  if (ms < 60000) return `${Math.round(ms / 1000)}s`
  return `${Math.round(ms / 60000)}m`
}

// 从后端 Config 解析为编辑器数据
export function fromScenarioConfig(scenarioId: number | string, config: any, fallbackName = '未命名场景'): ScenarioEditorData {
  const data: ScenarioEditorData = {
    uid: String(scenarioId),
    name: fallbackName,
    description: '',
    requests: [],
    variables: [],
    thresholds: { p95Ms: 0, p99Ms: 0, errorRate: 0, minRPS: 0 },
    load: { vus: 100, duration: '5m', rampUpType: 'linear', rampUpDuration: '30s', targetRPS: 0, thinkTimeMs: 0 },
    tags: [],
  }

  if (!config) return data

  if (config.load) {
    data.load = {
      vus: config.load.vus || 100,
      duration: config.load.duration || '5m',
      rampUpType: config.load.rampUp?.type || 'linear',
      rampUpDuration: config.load.rampUp?.duration || '30s',
      targetRPS: config.load.targetRPS || 0,
      thinkTimeMs: durationToMs(config.load.thinkTime),
    }
  }

  if (config.requests && Array.isArray(config.requests)) {
    data.requests = config.requests.map((r: any) => ({
      uid: generateUid(),
      name: r.name || 'Request',
      method: r.method || 'GET',
      url: r.url || '',
      headers: Object.entries(r.headers || {}).map(([k, v]) => ({
        uid: generateUid(),
        key: k,
        value: String(v),
        enabled: true,
      })),
      bodyType: r.bodyRaw ? 'raw' : (r.body ? 'json' : 'none'),
      bodyJson: r.body ? JSON.stringify(r.body, null, 2) : '{}',
      bodyRaw: r.bodyRaw || '',
      extractors: Object.entries(r.extractors || {}).map(([k, v]) => ({
        uid: generateUid(),
        name: k,
        jsonPath: String(v),
        enabled: true,
      })),
      assertions: {
        status: r.assertions?.status || [200],
        maxLatencyMs: r.assertions?.maxLatencyMs || 0,
        bodyContains: r.assertions?.bodyContains || [],
        enabled: !!r.assertions,
      },
      thinkTime: durationToMs(r.thinkTime),
      weight: r.weight || 1,
    }))
  }

  if (config.variables) {
    data.variables = Object.entries(config.variables).map(([k, v]) => ({
      uid: generateUid(),
      name: k,
      values: Array.isArray(v) ? v.map(String) : [String(v)],
      enabled: true,
    }))
  }

  if (config.thresholds) {
    data.thresholds = {
      p95Ms: config.thresholds.p95Ms || 0,
      p99Ms: config.thresholds.p99Ms || 0,
      errorRate: config.thresholds.errorRate || 0,
      minRPS: config.thresholds.minRPS || 0,
    }
  }

  return data
}

function durationToMs(d: string | undefined): number {
  if (!d) return 0
  // 简单解析:支持 "5m", "30s", "1h", "500ms"
  const match = /^(\d+(?:\.\d+)?)(ms|s|m|h)?$/.exec(d.trim())
  if (!match) return 0
  const num = parseFloat(match[1])
  const unit = match[2] || 's'
  switch (unit) {
    case 'ms': return Math.round(num)
    case 's': return Math.round(num * 1000)
    case 'm': return Math.round(num * 60000)
    case 'h': return Math.round(num * 3600000)
    default: return 0
  }
}

export function createEmptyEditorData(scenarioId?: number | string): ScenarioEditorData {
  return {
    uid: scenarioId ? String(scenarioId) : 'new',
    name: '未命名场景',
    description: '',
    requests: [createEmptyRequest('First Request')],
    variables: [],
    thresholds: { p95Ms: 0, p99Ms: 0, errorRate: 0, minRPS: 0 },
    load: { vus: 100, duration: '5m', rampUpType: 'linear', rampUpDuration: '30s', targetRPS: 0, thinkTimeMs: 0 },
    tags: [],
  }
}

File: internal/model/scenario.go
// Package model 定义 Pulse 的领域模型
package model

import (
	"database/sql/driver"
	"encoding/json"
	"errors"
	"time"
)

// ScenarioSource 场景来源
type ScenarioSource string

const (
	ScenarioSourceManual    ScenarioSource = "manual"
	ScenarioSourceHARImport ScenarioSource = "har_import"
	ScenarioSourceTemplate  ScenarioSource = "template"
)

// ScenarioStatus 场景状态
type ScenarioStatus string

const (
	ScenarioStatusDraft    ScenarioStatus = "draft"
	ScenarioStatusActive   ScenarioStatus = "active"
	ScenarioStatusArchived ScenarioStatus = "archived"
)

// Scenario 是压测场景的领域模型
type Scenario struct {
	ID          int64           `json:"id" gorm:"primaryKey;autoIncrement"`
	UUID        string          `json:"uuid" gorm:"uniqueIndex;not null"`
	ProjectID   int64           `json:"project_id" gorm:"index;not null"`
	Name        string          `json:"name" gorm:"not null;size:128"`
	Description string          `json:"description" gorm:"type:text"`
	Config      ScenarioConfig  `json:"config" gorm:"type:text;not null"` // JSON 序列化
	Source      ScenarioSource  `json:"source" gorm:"size:32;default:'manual'"`
	Status      ScenarioStatus  `json:"status" gorm:"size:32;default:'draft'"`
	Tags        StringArray     `json:"tags" gorm:"type:text"` // JSON 数组
	Version     int             `json:"version" gorm:"default:1"`
	CreatedBy   int64           `json:"created_by" gorm:"index"`
	CreatedAt   time.Time       `json:"created_at" gorm:"autoIncrement"`
	UpdatedAt   time.Time       `json:"updated_at" gorm:"autoUpdateTime"`
}

// TableName 指定 GORM 表名
func (Scenario) TableName() string {
	return "scenarios"
}

// ScenarioConfig 是场景配置的完整结构
// 存储为 JSON,可灵活支持未来的扩展
type ScenarioConfig struct {
	// 压测配置
	Load LoadConfig `json:"load"`

	// 请求列表
	Requests []RequestConfig `json:"requests"`

	// 变量定义(参数化)
	Variables map[string]VariableValue `json:"variables,omitempty"`

	// 阈值定义
	Thresholds *ThresholdConfig `json:"thresholds,omitempty"`
}

// LoadConfig 压测负载配置
type LoadConfig struct {
	VUs          int          `json:"vus" binding:"required,min=1"`
	Duration     string       `json:"duration" binding:"required"` // e.g. "5m"
	RampUp       *RampUpConfig `json:"rampUp,omitempty"`
	TargetRPS    int          `json:"targetRPS,omitempty"`    // 0 = 不限
	ThinkTime    string       `json:"thinkTime,omitempty"`    // 默认 think time
	Iterations   int          `json:"iterations,omitempty"`  // 0 = 直到 duration 结束
}

// RampUpConfig 渐进式启动配置
type RampUpConfig struct {
	Type     string `json:"type"`     // linear / step / wave
	Duration string `json:"duration"` // e.g. "30s"
	Steps    int    `json:"steps,omitempty"` // for step type
}

// RequestConfig 单个 HTTP 请求配置
type RequestConfig struct {
	Name       string                 `json:"name" binding:"required"`
	Method     string                 `json:"method" binding:"required,oneof=GET POST PUT DELETE PATCH"`
	URL        string                 `json:"url" binding:"required"`
	Headers    map[string]string      `json:"headers,omitempty"`
	Body       map[string]interface{} `json:"body,omitempty"`
	BodyRaw    string                 `json:"bodyRaw,omitempty"` // 用于非 JSON body
	Extractors map[string]string      `json:"extractors,omitempty"` // JSONPath
	Assertions *AssertionConfig       `json:"assertions,omitempty"`
	ThinkTime  string                 `json:"thinkTime,omitempty"`
	Weight     int                    `json:"weight,omitempty"` // 多请求场景下的权重
}

// AssertionConfig 断言配置
type AssertionConfig struct {
	Status       []int    `json:"status,omitempty"`        // 期望的状态码列表
	MaxLatencyMs int      `json:"maxLatencyMs,omitempty"`  // 最大 P95 延迟
	BodyContains []string `json:"bodyContains,omitempty"`  // body 必须包含的字符串
	HeaderCheck  map[string]string `json:"headerCheck,omitempty"` // header 必须包含的键值
}

// VariableValue 是单个变量的定义
// 可以是单一值(标量)或多个值(数组,数据驱动测试)
type VariableValue struct {
	Value interface{} `json:"value"`
}

// ThresholdConfig 阈值配置
type ThresholdConfig struct {
	P95Ms        int     `json:"p95Ms,omitempty"`
	P99Ms        int     `json:"p99Ms,omitempty"`
	ErrorRate    float64 `json:"errorRate,omitempty"` // 0.01 = 1%
	MinRPS       int     `json:"minRPS,omitempty"`
	MaxAvgLatencyMs int  `json:"maxAvgLatencyMs,omitempty"`
}

// ============================================
// 自定义 GORM 数据类型
// ============================================

// ScenarioConfig 实现 driver.Valuer 和 sql.Scanner
// 用于在 SQLite 中存储为 TEXT(JSON)
func (c *ScenarioConfig) Value() (driver.Value, error) {
	if c == nil {
		return nil, nil
	}
	return json.Marshal(c)
}

// Scan 实现 sql.Scanner
func (c *ScenarioConfig) Scan(value interface{}) error {
	if value == nil {
		*c = ScenarioConfig{}
		return nil
	}
	var data []byte
	switch v := value.(type) {
	case []byte:
		data = v
	case string:
		data = []byte(v)
	default:
		return errors.New("ScenarioConfig: unsupported scan type")
	}
	return json.Unmarshal(data, c)
}

// StringArray 是 JSON 序列化的字符串数组
type StringArray []string

func (a *StringArray) Value() (driver.Value, error) {
	if a == nil {
		return "[]", nil
	}
	return json.Marshal(a)
}

func (a *StringArray) Scan(value interface{}) error {
	if value == nil {
		*a = StringArray{}
		return nil
	}
	var data []byte
	switch v := value.(type) {
	case []byte:
		data = v
	case string:
		data = []byte(v)
	default:
		return errors.New("StringArray: unsupported scan type")
	}
	return json.Unmarshal(data, a)
}